cpp/include/cudf/dfa/superstate.hpp

```diff
@@ -0,0 +1,224 @@
+ *
+ * @tparam State state which will be tracked - up to 16 unique values. usually an enum class.
+ * @tparam Instruction type which can be added to a state to transition in to a new state.
+ * @tparam N total number of states representable by State, up to 16 unique states.
+ */
+template <typename State, typename Instruction, uint8_t N>
+struct superstate {
+ private:
+  static const uint8_t BITS = superstate_policy<N>::BITS;
+  static const uint8_t MASK = superstate_policy<N>::MASK;
+
+  using Data = typename superstate_policy<N>::Data;
+
+  Data _data;
+
+ public:
+  /**
+   * @brief creates a superstate which represents all possible states and applied transitions
+   */
+  inline constexpr superstate() : _data(0)
+  {
+    for (auto i = 0; i < N; i++) { _data |= i << (i * BITS); }
+  }
+
+  /**
+   * @brief creates a superstate which represents only a single state
+   */
+  inline constexpr superstate(State state) : _data(0)
+  {
+    for (auto i = 0; i < N; i++) { _data |= static_cast<Data>(state) << (i * BITS); }
+  }
+
+  /**
+   * @brief create a superstate with a specific underlying value
+   * @param data the underlying value
+   */
+  inline constexpr superstate(Data data) : _data(data) {}
+
+  inline constexpr Data data() const { return _data; }
+
+  /**
+   * @brief converts to a single state using the least significant bits of the underlying data
+   *
+   * @return State a state created from the least significant bits of the underlying data
+   */
+  inline constexpr operator State() const { return static_cast<State>(_data & MASK); }
+
+  /**
+   * @brief transitions each underlying state using the provided instruction
+   *
+   * @note the + operator is called N times, once for each state.
+   *
+   * @param rhs the instruction used to transition all states
+   * @return the result of all transitions
+   */
+  inline constexpr superstate operator+(Instruction rhs) const
+  {
+    superstate result(0);
+    for (auto i = 0; i < N; i++) { result.set<false>(i, get(i) + rhs); }
+    return result;
+  }
+
+  /**
+   * @brief concatenates all underlying states and transitions in to a new superstate
+   *
+   * @note this operator is associative (but non-commutative) and can be used in prefix sums / scans
+   *
+   * @param rhs a superstate representing a series of transitions
+   * @return constexpr superstate
+   */
+  inline constexpr superstate operator+(superstate rhs) const
+  {
+    superstate result(0);
+    for (auto i = 0; i < N; i++) { result.set<false>(i, rhs.get(get(i))); }
+    return result;
+  }
+
+  /**
+   * @brief compares the underlying data equality. all represented states must match exactly.
+   *
+   * @param rhs value to be compared
+   * @return true if all represented states match
+   * @return false if at least one represented state does not match
+   */
+  inline constexpr bool operator==(superstate rhs) const { return _data == rhs._data; }
```

Comment:
It's usually preferred to implement things like `operator==` and `operator+` as free functions unless there is a strong reason to do otherwise.